### PR TITLE
API and implementation of codefixes and code actions for Rascal itself

### PR DESCRIPTION
**Pull request by jurgenvinju**
* [x] lang::rascal::lsp::Actions module contains:
   * [x] Command definitions
   * [x] rascalCodeActions functions to detect code actions from a focus
   * [x] rascalExecutor for executing commands
* [x] factor common code for code actions from ParameterizedLSP in common class
* [x] introduce command execution to Rascal document service
* [x] introduce codeAction LSP message
   * [x] merge in quickfixes from Diagnostics
   * [x] call Actions module for more actions
* [x] testing manually 
* [x] adding test code
* [x] think of a way to expose `Command` and `CodeAction` definitions to rascal-core, such that the checker can annotate messages with quickfixes
* [x] try to remove more clones by factoring actions streaming code with function parameters.

https://github.com/user-attachments/assets/77d1a5a2-5109-4bee-a27b-63656d6b81dd

### CODE REVIEW

**Comment by jurgenvinju** [comment]
@DavyLandman @toinehartman this is still a draft but if you are interested and perhaps have feedback... that would be nice.

I'm trying to prepare us for the inevitable removed of `Tree@\loc`, and some quickfixes and code actions in the editor to help our users and ourselves with this would be really useful. The DSL part of this follow the exact same design on purpose.

**Comment by jurgenvinju** [comment]
* Typically the type-checker would attach CodeActions to error messages, for quick-fix purposes
* And for more general tools/refactorings/visualizations:
   * visualize import/extend graph
   * visualize call graph
   * convert curly body function to expression function
   * add parameter to all overloads
   * convert switch to overloaded function
   * ... etc ... the possibilities are endless :-)

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (toinehartman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:104]: Why does this directly call `applyDocumentEdits` while the top-level rewrite on line 79 returns an `action(..., kind = refactor())`?

**Review by toinehartman — COMMENTED** [review]
Nice improvement! I like how this allows to write transformations and other code actions relatively easily. I am curious how you'll tackle contribution code actions from the type checker!

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (jurgenvinju) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:104]: It's a demo of the possibilities. There is no semantic difference other than a different time of synchronization. The user does not see a difference. Both ways are applied in the editor.

Comment (jurgenvinju) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:104]: However, there are reasons for choosing a command over edits, typically if the amount of edits is very large or expensive to compute, then the action edits will block the creation of the actions menu. In that case it is always better to register a command that will compute the edits later.

Comment (jurgenvinju) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:104]: other reasons are if you need other side-effects next to the edits, then a command is useful. If you pick both, say an action with both edits and a command, then the edits will always be executed before the command is executed.

**Comment by jurgenvinju** [comment]
> Nice improvement! I like how this allows to write transformations and other code actions relatively easily. I am curious how you'll tackle contribution code actions from the type checker!

I was planning on learning that from you @toinehartman

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:110]: I'm not a fan of yet another evalutor getting added, this creates a bigger memory pressure. It also slows down initial startup, as they are all loaded in parallel when VS Code starts. 

Can't this one go into the `semanticEvaluator`? I mean, it's a balance, since `Summary` and especially `Rename` can also take quite a while, so it would mean the actions waiting in the queue.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:301]: whe here a `logger.catching` but not in the `parseCodeActions`?

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:344]: Can we instead of `null`  return an empty list?

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:287]: can we return some empty `IValue`? `Command::noop` I'm guessing? instead of null?

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (DavyLandman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:73]: not everyone wants to put a license header in their source code. Some licenses allow for only a license in the root of the directory. Can a user disable these in some way?

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (DavyLandman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:113]: nice one 👍🏼

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (DavyLandman) [inline rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:65]: we should only offer this if the header is more than 1 import?

**Review by DavyLandman — CHANGES_REQUESTED** [review]
I like how most of it looks 👍🏼 great reuse of the existing code-actions added the previous PR. My only concern is with the extra evaluator.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (jurgenvinju) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:65]: It shows all modules in the entire project always, so it is not that specific to the current module.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (jurgenvinju) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:73]: You can simply not choose to activate quickfix? If my memory does not fail me, the fact that licenses can only be in the root of the folder is less depending on the license and more on the locality and the scope of the license. American law is quite different from European law, and anglosaxon law from continental European law. 

In any case, the prudent lawyer would advice you to put it everywhere if you (a) don't want to be liable for the contents if someone copies an entire file without the project contents and (b) want the copier of the file to be liable to the conditions made in the license. 

In that sense this action is useful to anybody who wishes to be on the safe side.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:287]: I'll have a look. We can return `("result":false)` for backward compatibility.

Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:287]: But in reality this is dead since that's what `evaluateRascalCommand` does already.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:344]: check! Although this can never happen since `rascalCodeActions` returns the empty list by default.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:301]: right, because the failure would have occurred there already earlier. So we fail hard there as soon as possible, and softer here.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java
Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalLanguageServices.java:110]: I think that would be acceptible. Currently my machine is only a bit sooner to present the actions than to present the summary. And you are right there will be a large overlap between summary code and actions code eventually anyway, so leading to memory duplication that is unwarranted.

File: rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc
Comment (jurgenvinju) [reply rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc:73]: I've restricted the action to only appear if the cursor is on the QualifiedName part of the header (the module name). That way it is always within reach but not annoyingly so.

**Comment by jurgenvinju** [comment]
I think this is ready if you do to. Note that this does not test and quickfixes attached to messages from the checker, yet. That's because the checker does not produce them yet. I plan to add a test here anway, as soon as one of those quickfixes starts popping up by the checker. But that's after merging IMHO.

**Comment by jurgenvinju** [comment]
> I like how most of it looks 👍🏼 great reuse of the existing code-actions added the previous PR. My only concern is with the extra evaluator.

I fixed the evaluator but the PR still thinks this is open? Help, where do I click?

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:440]: I'm positive I've seen this code before, can we reduce the duplication? especially as it's easy to mess this up.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java
Comment (DavyLandman) [inline rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:464]: please move this calculation of the focus list into the completeable future, to move as much of the calculation to async task.

File: rascal-vscode-extension/src/test/vscode-suite/ide.test.ts
Comment (DavyLandman) [inline rascal-vscode-extension/src/test/vscode-suite/ide.test.ts:202]: we should refactor this into a function into the IDE class, with the text as an argument.

**Review by DavyLandman — APPROVED** [review]
Some small remarks after a second review. I've changed my vote to approve, such that if you've went through them, and dealt with them to your discretion, you can merge.

**Review by toinehartman — APPROVED** [review]

File: rascal-vscode-extension/src/test/vscode-suite/ide.test.ts
Comment (jurgenvinju) [reply rascal-vscode-extension/src/test/vscode-suite/ide.test.ts:202]: ok will do

**Comment by jurgenvinju** [comment]
@toinehartman right. Maybe we could make a Position to Position method like: `Position fromLSPpositionToRascalPosition(Position p)` . It is fairly critical code with weird constants and everything.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java
Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java:461]: this code was already called from the asynhronous task on lin 452. but I created yet another async wrapper anyway. Don't know if that helps.

File: rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/BaseWorkspaceService.java
Comment (jurgenvinju) [reply rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/BaseWorkspaceService.java:113]: Yes we are, because otherwise the LSP would break. In fact we were matching the wrong name here (and als passing the wrong name) and things got jumbled up between the two LSP servers.

**Comment by jurgenvinju** [comment]
@toinehartman @DavyLandman we now have single-point-of-change for translations from LSP startLine/column to Rascal's lines and columns with `Locations::toRascalPosition` applied everywhere. I've search for `+ 1`.

**Comment by toinehartman** [comment]
Thanks @jurgenvinju! Nice work. Looking forward to trying out the initial refactorings *and* use of this framework to add others.

**Pull request merged**
